### ISSUE TITLE
Rule Proposal: no-optional-eval-call

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#13496]
**Please describe what the rule should do:**

Disallow `eval?.(arg)` and suggest `eval(arg)` or `(0,eval)(arg)` to clarify.

There are some reasons:

- Each implementation behaves differently.
  - *indirect* eval: Firefox 79, GraalJS, XS (moddable), and engine262. (as per the current language spec.)
  - *direct* eval: Firefox 78, Chrome, Safari, and Node.js.
- As https://github.com/tc39/ecma262/issues/2062 mentioned, it's not clear if `eval?.(arg)` is direct eval or indirect eval. So a more clear notation is better.
  - People may expect that direct eval is *only* `eval(arg)` and all others are indirect eval. (the current language spec is so.)
  - People may expect that `eval?.(arg)` is exactly same as `eval(arg)` if `eval` is not nullish.

**What new ECMAScript feature does this rule relate to?**

Optional chaining.

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
//✘ BAD
eval?.(arg);

//✓ GOOD
eval(arg);
(0, eval)(arg);
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This rule is about the language spec and warns it as it may behave differently for each runtime.

**Are you willing to submit a pull request to implement this rule?**

Yes.

**Comment by mysticatea** [comment]
In the TC39 meeting July 2020, there is a consensus that `eval?.(arg)` is indirect eval. Currently, most runtime (Firefox Stable, Chrome, Safari, Node.js, TypeScript, and Babel) are direct eval as spec violation, but it should be fixed soon.

Therefore, this may be just stylistic in the near future.

**Comment by mdjermanovic** [comment]
Maybe we could choose a more generic name for this rule, so that it might be able to check other call syntaxes like `arg |> eval` in the future.

**Comment by nzakas** [comment]
This has been open for three years and doesn't fit the criteria for new rules, so closing.
